File: sdk/python/novis_sdk.py
"""
NOVIS SDK for Python

Gasless transactions for humans and AI agents on Base.

Version: 1.0.0
License: MIT
"""

import json
import time
from typing import Optional, Dict, Any, List
from dataclasses import dataclass
from decimal import Decimal

import requests
from web3 import Web3
from eth_account import Account
from eth_account.messages import encode_typed_data

# =============================================================================
# CONSTANTS
# =============================================================================

ADDRESSES = {
    "NOVIS_TOKEN": "0x1fb5e1C0c3DEc8da595E531b31C7B30c540E6B85",
    "VAULT": "0xA3D771bF986174D9cf9C85072cCD11cb72A694d4",
    "STRATEGY": "0x064E4586b7C63777BDC98A4776D3f78A93C0B752",
    "FACTORY": "0x4b84E3a0D640c9139426f55204Fb34dB9B1123EA",
    "DEX_POOL": "0xA0af1C990433102EFb08D78E060Ab05E6874ca69",
    "TREASURY": "0x4709280aef7A496EA84e72dB3CAbAd5e324d593e",
    "USDC": "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
    "ENTRYPOINT": "0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789",
    "RELAYER_API": "https://novis-relayer-production.up.railway.app",
    "RPC_URL": "https://mainnet.base.org",
    "CHAIN_ID": 8453
}

NOVIS_ABI = [
    {"name": "name", "type": "function", "inputs": [], "outputs": [{"type": "string"}], "stateMutability": "view"},
    {"name": "symbol", "type": "function", "inputs": [], "outputs": [{"type": "string"}], "stateMutability": "view"},
    {"name": "decimals", "type": "function", "inputs": [], "outputs": [{"type": "uint8"}], "stateMutability": "view"},
    {"name": "totalSupply", "type": "function", "inputs": [], "outputs": [{"type": "uint256"}], "stateMutability": "view"},
    {"name": "balanceOf", "type": "function", "inputs": [{"name": "account", "type": "address"}], "outputs": [{"type": "uint256"}], "stateMutability": "view"},
    {"name": "transfer", "type": "function", "inputs": [{"name": "to", "type": "address"}, {"name": "amount", "type": "uint256"}], "outputs": [{"type": "bool"}], "stateMutability": "nonpayable"},
    {"name": "feeThreshold", "type": "function", "inputs": [], "outputs": [{"type": "uint256"}], "stateMutability": "view"},
    {"name": "feePercentageBps", "type": "function", "inputs": [], "outputs": [{"type": "uint16"}], "stateMutability": "view"},
    {"name": "totalFeesCollected", "type": "function", "inputs": [], "outputs": [{"type": "uint256"}], "stateMutability": "view"},
    {"name": "totalMetaTxRelayed", "type": "function", "inputs": [], "outputs": [{"type": "uint256"}], "stateMutability": "view"},
    {"name": "calculateTransferFee", "type": "function", "inputs": [{"name": "from", "type": "address"}, {"name": "to", "type": "address"}, {"name": "amount", "type": "uint256"}], "outputs": [{"type": "uint256"}, {"type": "uint256"}], "stateMutability": "view"},
]

VAULT_ABI = [
    {"name": "deposit", "type": "function", "inputs": [{"name": "amount", "type": "uint256"}], "outputs": [{"type": "uint256"}], "stateMutability": "nonpayable"},
    {"name": "redeem", "type": "function", "inputs": [{"name": "amount", "type": "uint256"}], "outputs": [{"type": "uint256"}], "stateMutability": "nonpayable"},
    {"name": "totalAssets", "type": "function", "inputs": [], "outputs": [{"type": "uint256"}], "stateMutability": "view"},
    {"name": "backingRatioBps", "type": "function", "inputs": [], "outputs": [{"type": "uint256"}], "stateMutability": "view"},
]

FACTORY_ABI = [
    {"name": "createAccount", "type": "function", "inputs": [{"name": "owner", "type": "address"}, {"name": "dailyLimit", "type": "uint256"}, {"name": "salt", "type": "bytes32"}], "outputs": [{"type": "address"}], "stateMutability": "nonpayable"},
    {"name": "accountCount", "type": "function", "inputs": [], "outputs": [{"type": "uint256"}], "stateMutability": "view"},
    {"name": "accounts", "type": "function", "inputs": [{"name": "index", "type": "uint256"}], "outputs": [{"type": "address"}], "stateMutability": "view"},
    {"name": "getAccountsByOwner", "type": "function", "inputs": [{"name": "owner", "type": "address"}], "outputs": [{"type": "address[]"}], "stateMutability": "view"},
]

USDC_ABI = [
    {"name": "balanceOf", "type": "function", "inputs": [{"name": "account", "type": "address"}], "outputs": [{"type": "uint256"}], "stateMutability": "view"},
    {"name": "approve", "type": "function", "inputs": [{"name": "spender", "type": "address"}, {"name": "amount", "type": "uint256"}], "outputs": [{"type": "bool"}], "stateMutability": "nonpayable"},
    {"name": "allowance", "type": "function", "inputs": [{"name": "owner", "type": "address"}, {"name": "spender", "type": "address"}], "outputs": [{"type": "uint256"}], "stateMutability": "view"},
]


# =============================================================================
# DATA CLASSES
# =============================================================================

@dataclass
class TransferResult:
    success: bool
    tx_hash: str
    block_number: int
    from_address: str
    to_address: str
    amount: str
    explorer_url: str


@dataclass
class FeeInfo:
    amount: str
    amount_wei: str
    fee: str
    fee_wei: str
    net_amount: str
    net_amount_wei: str
    fee_percent: str


@dataclass
class ProtocolStats:
    total_supply: str
    total_assets: str
    backing_ratio_bps: int
    backing_ratio_percent: str
    total_fees_collected: str
    total_meta_tx_relayed: int
    fee_threshold: str
    fee_percentage: str


# =============================================================================
# NOVIS CLIENT
# =============================================================================

class NOVISClient:
    """
    NOVIS SDK Client for Python
    
    Example:
        client = NOVISClient(private_key="0x...")
        result = client.transfer("0xRecipient...", "10.0")
        print(result.tx_hash)
    """
    
    def __init__(
        self,
        private_key: str,
        rpc_url: str = None,
        relayer_url: str = None
    ):
        """
        Initialize NOVIS client
        
        Args:
            private_key: Wallet private key (with or without 0x prefix)
            rpc_url: Optional custom RPC URL
            relayer_url: Optional custom relayer URL
        """
        self.rpc_url = rpc_url or ADDRESSES["RPC_URL"]
        self.relayer_url = relayer_url or ADDRESSES["RELAYER_API"]
        
        self.w3 = Web3(Web3.HTTPProvider(self.rpc_url))
        self.account = Account.from_key(private_key)
        
        self.novis = self.w3.eth.contract(
            address=Web3.to_checksum_address(ADDRESSES["NOVIS_TOKEN"]),
            abi=NOVIS_ABI
        )
        self.vault = self.w3.eth.contract(
            address=Web3.to_checksum_address(ADDRESSES["VAULT"]),
            abi=VAULT_ABI
        )
        self.factory = self.w3.eth.contract(
            address=Web3.to_checksum_address(ADDRESSES["FACTORY"]),
            abi=FACTORY_ABI
        )
        self.usdc = self.w3.eth.contract(
            address=Web3.to_checksum_address(ADDRESSES["USDC"]),
            abi=USDC_ABI
        )
    
    @property
    def address(self) -> str:
        """Get wallet address"""
        return self.account.address
    
    # =========================================================================
    # BALANCE & INFO
    # =========================================================================
    
    def get_balance(self, address: str = None) -> str:
        """
        Get NOVIS balance
        
        Args:
            address: Address to check (defaults to wallet)
            
        Returns:
            Balance in NOVIS (human readable)
        """
        addr = Web3.to_checksum_address(address or self.address)
        balance = self.novis.functions.balanceOf(addr).call()
        return str(Web3.from_wei(balance, 'ether'))
    
    def get_usdc_balance(self, address: str = None) -> str:
        """Get USDC balance (6 decimals)"""
        addr = Web3.to_checksum_address(address or self.address)
        balance = self.usdc.functions.balanceOf(addr).call()
        return str(Decimal(balance) / Decimal(10**6))
    
    def get_eth_balance(self, address: str = None) -> str:
        """Get ETH balance"""
        addr = Web3.to_checksum_address(address or self.address)
        balance = self.w3.eth.get_balance(addr)
        return str(Web3.from_wei(balance, 'ether'))
    
    def get_protocol_stats(self) -> ProtocolStats:
        """Get protocol statistics"""
        total_supply = self.novis.functions.totalSupply().call()
        total_assets = self.vault.functions.totalAssets().call()
        backing_ratio = self.vault.functions.backingRatioBps().call()
        total_fees = self.novis.functions.totalFeesCollected().call()
        total_meta_tx = self.novis.functions.totalMetaTxRelayed().call()
        fee_threshold = self.novis.functions.feeThreshold().call()
        fee_bps = self.novis.functions.feePercentageBps().call()
        
        return ProtocolStats(
            total_supply=str(Web3.from_wei(total_supply, 'ether')),
            total_assets=str(Decimal(total_assets) / Decimal(10**6)),
            backing_ratio_bps=backing_ratio,
            backing_ratio_percent=f"{backing_ratio / 100:.2f}%",
            total_fees_collected=str(Web3.from_wei(total_fees, 'ether')),
            total_meta_tx_relayed=total_meta_tx,
            fee_threshold=str(Web3.from_wei(fee_threshold, 'ether')),
            fee_percentage=f"{fee_bps / 100:.2f}%"
        )
    
    # =========================================================================
    # GASLESS TRANSFER
    # =========================================================================
    
    def transfer(self, to: str, amount: str) -> TransferResult:
        """
        Send NOVIS gaslessly via meta-transaction
        
        Args:
            to: Recipient address
            amount: Amount in NOVIS (e.g., "10.0")
            
        Returns:
            TransferResult with transaction details
        """
        to = Web3.to_checksum_address(to)
        amount_wei = Web3.to_wei(Decimal(amount), 'ether')
        
        # 1. Get nonce
        nonce_res = requests.get(f"{self.relayer_url}/nonce/{self.address}")
        nonce_res.raise_for_status()
        nonce = int(nonce_res.json()["nonce"])
        
        # 2. Get domain
        domain_res = requests.get(f"{self.relayer_url}/domain")
        domain_res.raise_for_status()
        domain_data = domain_res.json()
        
        # 3. Build EIP-712 typed data
        deadline = int(time.time()) + 3600  # 1 hour
        
        typed_data = {
            "types": {
                "EIP712Domain": [
                    {"name": "name", "type": "string"},
                    {"name": "version", "type": "string"},
                    {"name": "chainId", "type": "uint256"},
                    {"name": "verifyingContract", "type": "address"}
                ],
                "MetaTransfer": [
                    {"name": "from", "type": "address"},
                    {"name": "to", "type": "address"},
                    {"name": "amount", "type": "uint256"},
                    {"name": "nonce", "type": "uint256"},
                    {"name": "deadline", "type": "uint256"}
                ]
            },
            "primaryType": "MetaTransfer",
            "domain": {
                "name": domain_data["name"],
                "version": domain_data["version"],
                "chainId": int(domain_data["chainId"]),
                "verifyingContract": domain_data["verifyingContract"]
            },
            "message": {
                "from": self.address,
                "to": to,
                "amount": amount_wei,
                "nonce": nonce,
                "deadline": deadline
            }
        }
        
        # 4. Sign
        encoded = encode_typed_data(full_message=typed_data)
        signed = self.account.sign_message(encoded)
        signature = signed.signature.hex()
        
        # 5. Relay
        relay_res = requests.post(
            f"{self.relayer_url}/relay",
            json={
                "from": self.address,
                "to": to,
                "amount": str(amount_wei),
                "deadline": str(deadline),
                "signature": signature
            }
        )
        result = relay_res.json()
        
        if not result.get("success"):
            raise Exception(result.get("error", "Relay failed"))
        
        return TransferResult(
            success=True,
            tx_hash=result["txHash"],
            block_number=result["blockNumber"],
            from_address=self.address,
            to_address=to,
            amount=amount,
            explorer_url=f"https://basescan.org/tx/{result['txHash']}"
        )
    
    def calculate_fee(self, to: str, amount: str) -> FeeInfo:
        """Calculate fee for a transfer"""
        to = Web3.to_checksum_address(to)
        amount_wei = Web3.to_wei(Decimal(amount), 'ether')
        
        fee, net = self.novis.functions.calculateTransferFee(
            self.address, to, amount_wei
        ).call()
        
        return FeeInfo(
            amount=amount,
            amount_wei=str(amount_wei),
            fee=str(Web3.from_wei(fee, 'ether')),
            fee_wei=str(fee),
            net_amount=str(Web3.from_wei(net, 'ether')),
            net_amount_wei=str(net),
            fee_percent="0.1%" if fee > 0 else "0%"
        )
    
    # =========================================================================
    # DIRECT TRANSFER
    # =========================================================================
    
    def transfer_direct(self, to: str, amount: str) -> Dict[str, Any]:
        """
        Send NOVIS directly (requires ETH for gas)
        
        Args:
            to: Recipient address
            amount: Amount in NOVIS
            
        Returns:
            Transaction result dict
        """
        to = Web3.to_checksum_address(to)
        amount_wei = Web3.to_wei(Decimal(amount), 'ether')
        
        tx = self.novis.functions.transfer(to, amount_wei).build_transaction({
            'from': self.address,
            'nonce': self.w3.eth.get_transaction_count(self.address),
            'gas': 100000,
            'gasPrice': self.w3.eth.gas_price
        })
        
        signed = self.account.sign_transaction(tx)
        tx_hash = self.w3.eth.send_raw_transaction(signed.raw_transaction)
        receipt = self.w3.eth.wait_for_transaction_receipt(tx_hash)
        
        return {
            "success": receipt.status == 1,
            "tx_hash": tx_hash.hex(),
            "block_number": receipt.blockNumber,
            "gas_used": receipt.gasUsed,
            "explorer_url": f"https://basescan.org/tx/{tx_hash.hex()}"
        }
    
    # =========================================================================
    # SMART ACCOUNTS
    # =========================================================================
    
    def create_smart_account(self, daily_limit: str) -> str:
        """
        Create a smart account for AI agent
        
        Args:
            daily_limit: Daily spending limit in NOVIS
            
        Returns:
            Smart account address
        """
        daily_limit_wei = Web3.to_wei(Decimal(daily_limit), 'ether')
        salt = Web3.keccak(text=str(time.time()))
        
        tx = self.factory.functions.createAccount(
            self.address,
            daily_limit_wei,
            salt
        ).build_transaction({
            'from': self.address,
            'nonce': self.w3.eth.get_transaction_count(self.address),
            'gas': 500000,
            'gasPrice': self.w3.eth.gas_price
        })
        
        signed = self.account.sign_transaction(tx)
        tx_hash = self.w3.eth.send_raw_transaction(signed.raw_transaction)
        self.w3.eth.wait_for_transaction_receipt(tx_hash)
        
        # Get account address
        count = self.factory.functions.accountCount().call()
        account_address = self.factory.functions.accounts(count - 1).call()
        
        return account_address
    
    def get_my_smart_accounts(self) -> List[str]:
        """Get all smart accounts owned by this wallet"""
        return self.factory.functions.getAccountsByOwner(self.address).call()


# =============================================================================
# UTILITY FUNCTIONS
# =============================================================================

def format_novis(wei: int) -> str:
    """Format wei to NOVIS"""
    return str(Web3.from_wei(wei, 'ether'))

def parse_novis(amount: str) -> int:
    """Parse NOVIS string to wei"""
    return Web3.to_wei(Decimal(amount), 'ether')

def format_usdc(wei: int) -> str:
    """Format wei to USDC (6 decimals)"""
    return str(Decimal(wei) / Decimal(10**6))

def parse_usdc(amount: str) -> int:
    """Parse USDC string to wei"""
    return int(Decimal(amount) * Decimal(10**6))

def is_valid_address(address: str) -> bool:
    """Check if address is valid"""
    return Web3.is_address(address)


# =============================================================================
# MAIN (for testing)
# =============================================================================

if __name__ == "__main__":
    import os
    
    pk = os.environ.get("PRIVATE_KEY")
    if not pk:
        print("Set PRIVATE_KEY environment variable")
        exit(1)
    
    client = NOVISClient(pk)
    print(f"Address: {client.address}")
    print(f"NOVIS Balance: {client.get_balance()}")
    print(f"ETH Balance: {client.get_eth_balance()}")
    
    stats = client.get_protocol_stats()
    print(f"\nProtocol Stats:")
    print(f"  Total Supply: {stats.total_supply} NOVIS")
    print(f"  Backing Ratio: {stats.backing_ratio_percent}")
    print(f"  Total Fees: {stats.total_fees_collected} NOVIS")
    print(f"  Meta-Txs: {stats.total_meta_tx_relayed}")

File: sdk/javascript/novis-sdk.js
/**
 * NOVIS SDK for JavaScript/TypeScript
 * 
 * Gasless transactions for humans and AI agents on Base.
 * 
 * @version 1.0.0
 * @license MIT
 */

import { ethers } from 'ethers';

// =============================================================================
// CONSTANTS
// =============================================================================

export const ADDRESSES = {
  NOVIS_TOKEN: "0x1fb5e1C0c3DEc8da595E531b31C7B30c540E6B85",
  VAULT: "0xA3D771bF986174D9cf9C85072cCD11cb72A694d4",
  STRATEGY: "0x064E4586b7C63777BDC98A4776D3f78A93C0B752",
  FACTORY: "0x4b84E3a0D640c9139426f55204Fb34dB9B1123EA",
  DEX_POOL: "0xA0af1C990433102EFb08D78E060Ab05E6874ca69",
  TREASURY: "0x4709280aef7A496EA84e72dB3CAbAd5e324d593e",
  USDC: "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
  ENTRYPOINT: "0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789",
  RELAYER_API: "https://novis-relayer-production.up.railway.app",
  RPC_URL: "https://mainnet.base.org",
  CHAIN_ID: 8453
};

const NOVIS_ABI = [
  "function name() view returns (string)",
  "function symbol() view returns (string)",
  "function decimals() view returns (uint8)",
  "function totalSupply() view returns (uint256)",
  "function balanceOf(address) view returns (uint256)",
  "function allowance(address,address) view returns (uint256)",
  "function transfer(address,uint256) returns (bool)",
  "function approve(address,uint256) returns (bool)",
  "function transferFrom(address,address,uint256) returns (bool)",
  "function feeThreshold() view returns (uint256)",
  "function feePercentageBps() view returns (uint16)",
  "function feesEnabled() view returns (bool)",
  "function calculateTransferFee(address,address,uint256) view returns (uint256,uint256)",
  "function getMetaTxNonce(address) view returns (uint256)",
  "function metaTransfer(address,address,uint256,uint256,bytes) returns (bool)",
  "function totalFeesCollected() view returns (uint256)",
  "function totalMetaTxRelayed() view returns (uint256)"
];

const VAULT_ABI = [
  "function deposit(uint256) returns (uint256)",
  "function redeem(uint256) returns (uint256)",
  "function totalAssets() view returns (uint256)",
  "function backingRatioBps() view returns (uint256)",
  "function paused() view returns (bool)"
];

const FACTORY_ABI = [
  "function createAccount(address,uint256,bytes32) returns (address)",
  "function accountCount() view returns (uint256)",
  "function accounts(uint256) view returns (address)",
  "function isAccount(address) view returns (bool)",
  "function getAccountsByOwner(address) view returns (address[])"
];

const USDC_ABI = [
  "function balanceOf(address) view returns (uint256)",
  "function approve(address,uint256) returns (bool)",
  "function allowance(address,address) view returns (uint256)"
];

// =============================================================================
// NOVIS CLIENT
// =============================================================================

export class NOVISClient {
  /**
   * Create a new NOVIS client
   * @param {string} privateKey - Wallet private key
   * @param {object} options - Optional configuration
   */
  constructor(privateKey, options = {}) {
    this.rpcUrl = options.rpcUrl || ADDRESSES.RPC_URL;
    this.relayerUrl = options.relayerUrl || ADDRESSES.RELAYER_API;
    
    this.provider = new ethers.JsonRpcProvider(this.rpcUrl);
    this.wallet = new ethers.Wallet(privateKey, this.provider);
    
    this.novis = new ethers.Contract(ADDRESSES.NOVIS_TOKEN, NOVIS_ABI, this.wallet);
    this.vault = new ethers.Contract(ADDRESSES.VAULT, VAULT_ABI, this.wallet);
    this.factory = new ethers.Contract(ADDRESSES.FACTORY, FACTORY_ABI, this.wallet);
    this.usdc = new ethers.Contract(ADDRESSES.USDC, USDC_ABI, this.wallet);
  }

  /**
   * Get wallet address
   * @returns {string} Address
   */
  get address() {
    return this.wallet.address;
  }

  // ===========================================================================
  // BALANCE & INFO
  // ===========================================================================

  /**
   * Get NOVIS balance
   * @param {string} address - Address to check (defaults to wallet)
   * @returns {Promise<string>} Balance in NOVIS (human readable)
   */
  async getBalance(address = null) {
    const addr = address || this.wallet.address;
    const balance = await this.novis.balanceOf(addr);
    return ethers.formatEther(balance);
  }

  /**
   * Get USDC balance
   * @param {string} address - Address to check (defaults to wallet)
   * @returns {Promise<string>} Balance in USDC (human readable, 6 decimals)
   */
  async getUSDCBalance(address = null) {
    const addr = address || this.wallet.address;
    const balance = await this.usdc.balanceOf(addr);
    return ethers.formatUnits(balance, 6);
  }

  /**
   * Get ETH balance
   * @param {string} address - Address to check (defaults to wallet)
   * @returns {Promise<string>} Balance in ETH
   */
  async getETHBalance(address = null) {
    const addr = address || this.wallet.address;
    const balance = await this.provider.getBalance(addr);
    return ethers.formatEther(balance);
  }

  /**
   * Get protocol stats
   * @returns {Promise<object>} Protocol statistics
   */
  async getProtocolStats() {
    const [
      totalSupply,
      totalAssets,
      backingRatio,
      totalFees,
      totalMetaTx,
      feeThreshold,
      feeBps
    ] = await Promise.all([
      this.novis.totalSupply(),
      this.vault.totalAssets(),
      this.vault.backingRatioBps(),
      this.novis.totalFeesCollected(),
      this.novis.totalMetaTxRelayed(),
      this.novis.feeThreshold(),
      this.novis.feePercentageBps()
    ]);

    return {
      totalSupply: ethers.formatEther(totalSupply),
      totalAssets: ethers.formatUnits(totalAssets, 6),
      backingRatioBps: Number(backingRatio),
      backingRatioPercent: (Number(backingRatio) / 100).toFixed(2) + '%',
      totalFeesCollected: ethers.formatEther(totalFees),
      totalMetaTxRelayed: Number(totalMetaTx),
      feeThreshold: ethers.formatEther(feeThreshold),
      feePercentage: (Number(feeBps) / 100).toFixed(2) + '%'
    };
  }

  // ===========================================================================
  // GASLESS TRANSFER (META-TRANSACTION)
  // ===========================================================================

  /**
   * Send NOVIS gaslessly via meta-transaction
   * @param {string} to - Recipient address
   * @param {string} amount - Amount in NOVIS (e.g., "10.0")
   * @returns {Promise<object>} Transaction result
   */
  async transfer(to, amount) {
    const amountWei = ethers.parseEther(amount);
    
    // 1. Get nonce
    const nonceRes = await fetch(`${this.relayerUrl}/nonce/${this.wallet.address}`);
    if (!nonceRes.ok) throw new Error('Failed to get nonce');
    const { nonce } = await nonceRes.json();
    
    // 2. Get domain
    const domainRes = await fetch(`${this.relayerUrl}/domain`);
    if (!domainRes.ok) throw new Error('Failed to get domain');
    const domainData = await domainRes.json();
    
    // 3. Build message
    const deadline = Math.floor(Date.now() / 1000) + 3600; // 1 hour
    
    const domain = {
      name: domainData.name,
      version: domainData.version,
      chainId: BigInt(domainData.chainId),
      verifyingContract: domainData.verifyingContract
    };
    
    const types = {
      MetaTransfer: [
        { name: "from", type: "address" },
        { name: "to", type: "address" },
        { name: "amount", type: "uint256" },
        { name: "nonce", type: "uint256" },
        { name: "deadline", type: "uint256" }
      ]
    };
    
    const message = {
      from: this.wallet.address,
      to: to,
      amount: amountWei,
      nonce: BigInt(nonce),
      deadline: BigInt(deadline)
    };
    
    // 4. Sign
    const signature = await this.wallet.signTypedData(domain, types, message);
    
    // 5. Relay
    const relayRes = await fetch(`${this.relayerUrl}/relay`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        from: this.wallet.address,
        to: to,
        amount: amountWei.toString(),
        deadline: deadline.toString(),
        signature: signature
      })
    });
    
    const result = await relayRes.json();
    
    if (!result.success) {
      throw new Error(result.error || 'Relay failed');
    }
    
    return {
      success: true,
      txHash: result.txHash,
      blockNumber: result.blockNumber,
      from: this.wallet.address,
      to: to,
      amount: amount,
      explorerUrl: `https://basescan.org/tx/${result.txHash}`
    };
  }

  /**
   * Calculate fee for a transfer
   * @param {string} to - Recipient address
   * @param {string} amount - Amount in NOVIS
   * @returns {Promise<object>} Fee information
   */
  async calculateFee(to, amount) {
    const amountWei = ethers.parseEther(amount);
    const [fee, netAmount] = await this.novis.calculateTransferFee(
      this.wallet.address,
      to,
      amountWei
    );
    
    return {
      amount: amount,
      amountWei: amountWei.toString(),
      fee: ethers.formatEther(fee),
      feeWei: fee.toString(),
      netAmount: ethers.formatEther(netAmount),
      netAmountWei: netAmount.toString(),
      feePercent: fee > 0n ? '0.1%' : '0%'
    };
  }

  // ===========================================================================
  // DIRECT TRANSFER (Requires ETH for gas)
  // ===========================================================================

  /**
   * Send NOVIS directly (requires ETH for gas)
   * @param {string} to - Recipient address
   * @param {string} amount - Amount in NOVIS
   * @returns {Promise<object>} Transaction result
   */
  async transferDirect(to, amount) {
    const amountWei = ethers.parseEther(amount);
    const tx = await this.novis.transfer(to, amountWei);
    const receipt = await tx.wait();
    
    return {
      success: receipt.status === 1,
      txHash: receipt.hash,
      blockNumber: receipt.blockNumber,
      gasUsed: receipt.gasUsed.toString(),
      explorerUrl: `https://basescan.org/tx/${receipt.hash}`
    };
  }

  // ===========================================================================
  // VAULT OPERATIONS
  // ===========================================================================

  /**
   * Deposit USDC to get NOVIS
   * @param {string} amount - Amount in USDC (e.g., "100")
   * @returns {Promise<object>} Transaction result
   */
  async deposit(amount) {
    const amountWei = ethers.parseUnits(amount, 6);
    
    // Check allowance
    const allowance = await this.usdc.allowance(this.wallet.address, ADDRESSES.VAULT);
    if (allowance < amountWei) {
      const approveTx = await this.usdc.approve(ADDRESSES.VAULT, amountWei);
      await approveTx.wait();
    }
    
    // Deposit
    const tx = await this.vault.deposit(amountWei);
    const receipt = await tx.wait();
    
    return {
      success: receipt.status === 1,
      txHash: receipt.hash,
      blockNumber: receipt.blockNumber,
      usdcDeposited: amount,
      explorerUrl: `https://basescan.org/tx/${receipt.hash}`
    };
  }

  /**
   * Redeem NOVIS for USDC
   * @param {string} amount - Amount in NOVIS (e.g., "100")
   * @returns {Promise<object>} Transaction result
   */
  async redeem(amount) {
    const amountWei = ethers.parseEther(amount);
    const tx = await this.vault.redeem(amountWei);
    const receipt = await tx.wait();
    
    return {
      success: receipt.status === 1,
      txHash: receipt.hash,
      blockNumber: receipt.blockNumber,
      novisRedeemed: amount,
      explorerUrl: `https://basescan.org/tx/${receipt.hash}`
    };
  }

  // ===========================================================================
  // SMART ACCOUNTS (For AI Agents)
  // ===========================================================================

  /**
   * Create a smart account for AI agent
   * @param {string} dailyLimit - Daily spending limit in NOVIS
   * @returns {Promise<string>} Smart account address
   */
  async createSmartAccount(dailyLimit) {
    const dailyLimitWei = ethers.parseEther(dailyLimit);
    const salt = ethers.keccak256(
      ethers.toUtf8Bytes(Date.now().toString() + Math.random())
    );
    
    const tx = await this.factory.createAccount(
      this.wallet.address,
      dailyLimitWei,
      salt
    );
    await tx.wait();
    
    // Get account address
    const count = await this.factory.accountCount();
    const accountAddress = await this.factory.accounts(count - 1n);
    
    return accountAddress;
  }

  /**
   * Get all smart accounts owned by this wallet
   * @returns {Promise<string[]>} Array of account addresses
   */
  async getMySmartAccounts() {
    return await this.factory.getAccountsByOwner(this.wallet.address);
  }

  /**
   * Fund a smart account with NOVIS
   * @param {string} accountAddress - Smart account address
   * @param {string} amount - Amount in NOVIS
   * @returns {Promise<object>} Transaction result
   */
  async fundSmartAccount(accountAddress, amount) {
    return await this.transferDirect(accountAddress, amount);
  }
}

// =============================================================================
// UTILITY FUNCTIONS
// =============================================================================

/**
 * Format NOVIS amount for display
 * @param {string|bigint} wei - Amount in wei
 * @returns {string} Formatted amount
 */
export function formatNOVIS(wei) {
  return ethers.formatEther(wei);
}

/**
 * Parse NOVIS amount from string
 * @param {string} amount - Amount in NOVIS
 * @returns {bigint} Amount in wei
 */
export function parseNOVIS(amount) {
  return ethers.parseEther(amount);
}

/**
 * Format USDC amount for display
 * @param {string|bigint} wei - Amount in wei (6 decimals)
 * @returns {string} Formatted amount
 */
export function formatUSDC(wei) {
  return ethers.formatUnits(wei, 6);
}

/**
 * Parse USDC amount from string
 * @param {string} amount - Amount in USDC
 * @returns {bigint} Amount in wei
 */
export function parseUSDC(amount) {
  return ethers.parseUnits(amount, 6);
}

/**
 * Check if an address is valid
 * @param {string} address - Ethereum address
 * @returns {boolean} True if valid
 */
export function isValidAddress(address) {
  return ethers.isAddress(address);
}

// =============================================================================
// DEFAULT EXPORT
// =============================================================================

export default NOVISClient;
